src/pykeen/nn/node_piece/anchor_search.py

```diff
@@ -222,6 +222,152 @@ def __call__(self, edge_index: numpy.ndarray, anchors: numpy.ndarray, k: int) ->
         return self.select(pool=pool, k=k)
 
 
+class SparseBFSSearcher(ScipySparseAnchorSearcher):
+    """Find closest anchors using :mod:`torch_sparse` on a GPU."""
+
+
+    @staticmethod
+    def create_adjacency(
+        edge_index: numpy.ndarray,
+    ) -> torch.tensor:
+        """
+        Create a sparse adjacency matrix (in the form of the edge list) from a given edge index.
+
+        :param edge_index: shape: (2, m)
+            the edge index
+
+        :return: shape: (2, 2m + n)
+            edge list with inverse edges and self-loops
+        """
+
+        # infer shape
+        num_entities = edge_index.max().item() + 1
+        edge_index = torch.tensor(edge_index, dtype=torch.long)
+
+        # symmetric + self-loops
+        # TODO what if the edge index already has inverse edges?
+        edge_list = torch.cat([
+            edge_index,
+            edge_index.flip(0),
+            torch.arange(num_entities).unsqueeze(0).repeat(2, 1)
+        ], dim=-1)
+
+        return edge_list
+
+    @staticmethod
+    def bfs(
+        anchors: numpy.ndarray,
+        edge_list: torch.tensor,
+        max_iter: int,
+        k: int,
+    ) -> numpy.ndarray:
+        """
+        Determine the candidate pool using breadth-first search.
+
+        :param anchors: shape: (a,)
+            the anchor node IDs
+        :param edge_list: shape: (2, n)
+            the edge list with symmetric edges and self-loops
+        :param max_iter:
+            the maximum number of hops to consider
+        :param k:
+            the minimum number of anchor nodes to reach
+
+        :return: shape: (n, a)
+            a boolean array indicating whether anchor $j$ is in the set of $k$ closest anchors for node $i$
+        """
+
+        try:
+            import torch_sparse
+            from torch_sparse import spmm
+        except ImportError as err:
+            raise ImportError(f"Requires `torch_sparse` to be installed.") from err
+
+        num_entities = edge_list.max().item() + 1
+        # for each entity, determine anchor pool by BFS
+        num_anchors = len(anchors)
+
+        device = resolve_device()
```

Comment:
maybe make this an attribute, and allow to explicitly pass a device in the `__init__`?

Reply:
if it's static, then it means there's no state, so just add an extra parameter to the static method that allows passing a device in, then when the static method is called from a stateful context, pass the self.device

Reply:
oh nice, done in https://github.com/pykeen/pykeen/pull/990/commits/36a257a511a5e89b361250cd2fddac2d5960507a